rct229/rules/section22/section22rule8.py

```diff
@@ -0,0 +1,141 @@
+from rct229.data.schema_enums import schema_enums
+from rct229.rule_engine.rule_base import RuleDefinitionBase
+from rct229.rule_engine.rule_list_indexed_base import RuleDefinitionListIndexedBase
+from rct229.rule_engine.user_baseline_proposed_vals import UserBaselineProposedVals
+from rct229.ruleset_functions.baseline_systems.baseline_system_util import HVAC_SYS
+from rct229.ruleset_functions.get_baseline_system_types import get_baseline_system_types
+from rct229.ruleset_functions.get_primary_secondary_loops_dict import (
+    get_primary_secondary_loops_dict,
+)
+from rct229.schema.config import ureg
+from rct229.utils.assertions import getattr_
+from rct229.utils.jsonpath_utils import find_all, find_one_with_field_value
+from rct229.utils.pint_utils import ZERO
+
+APPLICABLE_SYS_TYPES = [
+    HVAC_SYS.SYS_7,
+    HVAC_SYS.SYS_8,
+    HVAC_SYS.SYS_11_1,
+    HVAC_SYS.SYS_11_2,
+    HVAC_SYS.SYS_12,
+    HVAC_SYS.SYS_13,
+    HVAC_SYS.SYS_7B,
+    HVAC_SYS.SYS_8B,
+    HVAC_SYS.SYS_11_1B,
+    HVAC_SYS.SYS_12B,
+]
+PUMP_SPPED_CONTROL = schema_enums["PumpSpeedControlOptions"]
+FLUID_LOOP = schema_enums["FluidLoopOptions"]
+MIN_CHW_PRIMARY_LOOP_COOLING_CAPACITY = 300.0 * ureg("ton")
+
+
+class Section22Rule8(RuleDefinitionListIndexedBase):
+    """Rule 8 of ASHRAE 90.1-2019 Appendix G Section 22 (Chilled water loop)"""
+
+    def __init__(self):
+        super(Section22Rule8, self).__init__(
+            rmrs_used=UserBaselineProposedVals(False, True, False),
+            each_rule=Section22Rule8.ChillerFluidLoopRule(),
+            index_rmr="baseline",
+            id="22-8",
+            description="For Baseline chilled water system with cooling capacity of 300 tons or more, the secondary pump shall be modeled with variable-speed drives.",
+            list_path="ruleset_model_instances[0].fluid_loops[*]",
+        )
+
+    def is_applicable(self, context, data=None):
+        rmr_baseline = context.baseline
+        rmi_b = rmr_baseline["ruleset_model_instances"][0]
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+        # create a list contains all HVAC systems that are modeled in the rmi_b
+        available_type_lists = [
+            hvac_type
+            for hvac_type in baseline_system_types_dict.keys()
+            if len(baseline_system_types_dict[hvac_type]) > 0
+        ]
+
+        primary_secondary_loop_dictionary = get_primary_secondary_loops_dict(rmi_b)
+
+        return (
+            any(
+                [
+                    available_type in APPLICABLE_SYS_TYPES
+                    for available_type in available_type_lists
+                ]
+            )
+            and primary_secondary_loop_dictionary
+        )
+
+    def create_data(self, context, data):
+        rmr_baseline = context.baseline
+        rmi_b = rmr_baseline["ruleset_model_instances"][0]
+        loop_pump_dictionary = {
```

Comment:
In EnergyPlus, plant loops **usually** have one pump in the supply/demand side. ```usually`` means it can have more than one pump on the supply/demand side, but E+ users set one pump. In actual HVAC systems, there can be more than one pump in the loop.

Reply:
Sorry, I am a bit confused by this implementation. It seems this is a dict comprehension?
I agree we should stick with RDS implementation, which uses `lists` as the value for two reasons:
1. RMD is a general data schema for building energy models. If the schema supports multiple pumps in one loop, then we should take it into consideration.
2. Eplus can model multiple pumps at one side through headeredpump object, or interlock a pump with a chiller (A Branch object can have a pump and a chiller, the pump outlet shall be the chiller's inlet or vice versa). Modeling strategies could be tricky and often time it is default to general practices, but we need to constantly thinking about possibilities :).